design-documents/bit-precise-types.rst

```diff
@@ -0,0 +1,213 @@
+..
+   Copyright (c) 2023, Arm Limited and its affiliates.  All rights reserved.
+   CC-BY-SA-4.0 AND Apache-Patent-License
+   See LICENSE file for details
+
+Rationale Document for ABI related to the C23 _BitInt type.
+***********************************************************
+
+Preamble
+========
+
+Background
+----------
+
+This document describes the rationale behind the ABI choices made for using the
+bit-precise integral types defined in C2x.  These are ``_BitInt(N)`` and
+``unsigned _BitInt(N)``.  These are defined for integral ``N`` and each ``N`` is
+a different type.
+
+The proposal for these types can be found in following link.
+https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2763.pdf
+
+As the rationale mentioned, some applications have uses for a specific bit-width
+type.  In the case of writing C code which can be used to determine FPGA
+hardware these specific bit-width types can lead to large performance and space
+savings.
+
+From the perspective of the Arm ABI we have some trade-offs to determine.  We
+need to choose a representation for these objects in memory and in registers
+along with the size and alignment of the objects.  The main trade-offs we have
+identified in this case are on performance between different types of C-level
+operations, whether certain hardware-level atomic operations are possible, and
+general familiarity of programmers with the representation.
+
+For this particular type we are estimating that the use of ``_BitInt`` types
+will not be such that operations on these types are performance critical.
+
+There seem to be two different regimes for these types.  The "small" regime
+where bit-precise types could be stored in a single register, and the "large"
+regime where bit-precise types must span multiple registers.
```

Comment:
nitpick: maybe say general purpose register? Is the term "general
purpose register" used in other parts of the ABI docs?